### ISSUE TITLE
s3: rclone check and rclone copy s3->s3 slow and using too much memory

### ISSUE THREAD

**Issue by snailAlice** [rclone/rclone#6600]
#### What is the problem you are having with rclone?
When I execute the rclone check command, I find that the execution of the path containing 5744 files is too slow, it seems to be stuck somewhere, and consumes 10+G of memory, the entire bucket size is more than 1 T, and check this folder There are more than 1000 subfolders under it.
<img width="869" alt="image" src="https://user-images.githubusercontent.com/17794813/205533691-d19899ab-4de8-438d-82b6-b3aa87e7bdc7.png">

#### What is your rclone version (output from `rclone version`)
rclone-v1.60.1

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
CentOS Linux release 7.5.1804 (Core)

#### Which cloud storage system are you using? (e.g. Google Drive)

Alibaba and Minio

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
./rclone check oss-source:gz-gov-open-doc/gz-old-site/attachment/201512/ minio-target:gz-gov-open-doc/gz-old-site/attachment/201512/ --missing-on-dst 201608.txt --checkers 64 -vv

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by Masamune3210** [comment]
Rclone seems to have some pretty hard inefficiencies when it comes to folder transversal in general, at least on gdrive, not sure if that expands to other remotes or not though

**Comment by Masamune3210** [comment]
Not sure if inefficiencies is the right word, more like just slow points

**Comment by snailAlice** [comment]
> 一般而言，Rclone 在文件夹遍历方面似乎存在一些非常低效的问题，至少在 gdrive 上是这样，但不确定是否扩展到其他遥控器

In this case, how should I optimize it? Write a script to get the list of folders and then perform check and copy operations on it. Otherwise, I feel that its execution is too slow. I don’t know if you There is a better solution to this problem, best wishes.

**Comment by snailAlice** [comment]
> Rclone seems to have some pretty hard inefficiencies when it comes to folder transversal in general, at least on gdrive, not sure if that expands to other remotes or not though
In this case, how should I optimize it? Write a script to get the list of folders and then perform check and copy operations on it. Otherwise, I feel that its execution is too slow. I don’t know if you There is a better solution to this problem, best wishes.

**Comment by snailAlice** [comment]
> Rclone seems to have some pretty hard inefficiencies when it comes to folder transversal in general, at least on gdrive, not sure if that expands to other remotes or not though



In this case, how should I optimize it? Write a script to get the list of folders and then perform check and copy operations on it. Otherwise, I feel that its execution is too slow. I don’t know if you There is a better solution to this problem, best wishes.

**Comment by ncw** [comment]
> ./rclone check oss-source:gz-gov-open-doc/gz-old-site/attachment/201512/ minio-target:gz-gov-open-doc/gz-old-site/attachment/201512/ --missing-on-dst 201608.txt --checkers 64 -vv

This command looks OK to me.

`rclone check` on s3->s3 should be very quick.

> When I execute the rclone check command, I find that the execution of the path containing 5744 files is too slow, it seems to be stuck somewhere, and consumes 10+G of memory,

You are using `--checkers 64` which means rclone is loading 64 directories into memory at once. How many entries might each directory have? If it has millions then that would explain the 10G memory usage. If it doesn't then it sounds like there is a problem.

Can you run your command with `--rc` then collect a memory profile for me: https://rclone.org/rc/#debugging-memory-use

The most useful profile for me would be the one with `go tool pprof -svg http://localhost:5572/debug/pprof/heap` which will create an SVG which you can attach here.

This could also be a bug in one of the providers - if the directory listing is looping for some reason (this has happened before).

Have you got a particular directory that rclone is failing on?

If so can you try `rclone lsf --disable ListR` on that directory and see if it finishes properly? If not then run with `-vv --dump bodies --log-file rclone.log` and post the log.

So this is probably a bug I'd say, but whether it is in rclone, minio or Alibaba I can't say just yet!

**Comment by snailAlice** [comment]
svg file
[pprof.rclone.015.pb.gz](https://github.com/rclone/rclone/files/10154147/pprof.rclone.015.pb.gz)

**Comment by ncw** [comment]
Thank you for the logs.

From your log it is obvious what is happening - rclone is stuck in a directory loop

    GET /?delimiter=%2F&encoding-type=url&list-type=2&max-keys=1000&prefix=gz-old-site%2Fattachment%2F201511%2F HTTP/1.1

I decoded the XML here

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
  <Name>gz-gov-open-doc</Name>
  <Prefix>gz-old-site%2Fattachment%2F201511%2F</Prefix>
  <Marker>
  </Marker>
  <MaxKeys>1000</MaxKeys>
  <Delimiter>%2F</Delimiter>
  <EncodingType>url</EncodingType>
  <IsTruncated>true</IsTruncated>
  <NextMarker>gz-old-site%2Fattachment%2F201511%2Fxybgn%2F</NextMarker>
```

The `NextMarker` is set to `gz-old-site%2Fattachment%2F201511%2Fxybgn%2F` but rclone fetches `gz-old-site%2Fattachment%2F201511%2F` in the next response for some reason...

Rclone clearly is asking for a v2 listing type `list-type=2` but the response contains no `ContinuationToken` (see the docs here: https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html) - that is what is causing the loop. In fact the response is clearly a v1 listing type since it contains `NextMarker`.

Can you try this please

[v1.61.0-beta.6591.32a731543.fix-6600-s3-listing](https://beta.rclone.org/branch/fix-6600-s3-listing/v1.61.0-beta.6591.32a731543.fix-6600-s3-listing/) on branch [fix-6600-s3-listing](https://github.com/rclone/rclone/tree/fix-6600-s3-listing) (uploaded in 15-30 mins)

It should error out with

> s3 protocol error: received listing v2 with IsTruncated set and no ContinuationToken. Should you be using `--s3-list-version 1`?

Assuming it does, can you then try  `--s3-list-version 1`

I just did a test with Alibaba OSS myself, and it looks like it doesn't properly support v2 listing, so rclone's default is incorrect.

Here is another beta (with the previous change) - this should work without the `--s3-list-version` parameter

[v1.61.0-beta.6592.acfa8ae87.fix-6600-s3-listing](https://beta.rclone.org/branch/fix-6600-s3-listing/v1.61.0-beta.6592.acfa8ae87.fix-6600-s3-listing/) on branch [fix-6600-s3-listing](https://github.com/rclone/rclone/tree/fix-6600-s3-listing) (uploaded in 15-30 mins)

In summary I've added two fixes to rclone

1. detect a List v1 server not supporting List v2 and return an error instead of a directory loop
2. Change the default for `provider = Alibaba` to list v1
 
Fix 1. will stop the loops happening to the next unlucky user and 2) will fix it for you hopefully.

**Comment by snailAlice** [comment]
Thank you very much for your help. The beta version of rclone with the --s3-list-version 1 parameter works very well. I thought about it, maybe the OSS version used was built privately, and the use of Alibaba Cloud OSS is very easy. The current problem only appeared in the previous version. In addition, I would like to ask you whether this beta version needs to add additional parameters when executing the copy command, so as to avoid the situation of falling into a loop when the copy command is executed.once again express my sincere thanks.

**Comment by snailAlice** [comment]
Thank you very much for your help. The beta version of rclone with the --s3-list-version 1 parameter works very well. I thought about it, maybe the OSS version used was built privately, and the use of Alibaba Cloud OSS is very easy. The current problem only appeared in the previous version.once again express my sincere thanks.

**Comment by snailAlice** [comment]
Thank you very much for your help. The beta version of rclone works very well. I thought about it, maybe the OSS version used was built privately, and the use of Alibaba Cloud OSS is very easy. The current problem only appeared in the previous version.once again express my sincere thanks.

**Comment by ncw** [comment]
Thank you for testing this.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.61



> I thought about it, maybe the OSS version used was built privately, and the use of Alibaba Cloud OSS is very easy. The current problem only appeared in the previous version.once again express my sincere thanks.

It looks like you are using a private version of OSS but I checked that the public cloud OSS behaves the same.

**Comment by ncw** [comment]
@snailAlice I realised I made a mistake in my analysis of Alibaba OSS and it **does** support list versions v2.

So I'm going to re-enable it for Alibaba OSS.

This means you will have to use `--s3-list-version 1` or put `list_version 1` in your config file to use your private OSS.

**Comment by snailAlice** [comment]
Hello, I recently synchronized another cloud object storage with rclone and encountered the above problems. It seems that the beta version you gave last time does not work. There is no response in a certain folder of lsf. The attachment is the corresponding debug log. , but also trouble you to help me look at this problem.
[rclone.log.tar.gz](https://github.com/rclone/rclone/files/10234312/rclone.log.tar.gz)

**Comment by snailAlice** [comment]
> 
> This means you will have to use `--s3-list-version 1` or put `list_version 1` in your config file to use your private OSS.

Yes, I did this, and added the use --s3-list-version 1 parameter when executing.

**Comment by snailAlice** [comment]
@ncw Ok, I'll do it with a higher configuration server.
